pywemo/ssdp.py

```diff
@@ -336,16 +366,137 @@ def scan(
+    """
+
+    def __init__(self, callback_port: int):
+        """Create a server that will respond to WeMo discovery requests.
+
+        Args:
+            callback_port: The port for the SubscriptionRegistry HTTP server.
+        """
+        self.callback_port = callback_port
+        self._thread = None
+        self._exit = threading.Event()
+        self._thread_exception = None
+        self._notify_enabled = True  # For tests.
+
+    def send_notify(self) -> None:
+        """Send a UPnP NOTIFY message containing the virtual device URL."""
+        ssdp_target = (MULTICAST_GROUP, MULTICAST_PORT)
+        for addr in interface_addresses():  # Send on all interfaces.
+            callback_addr = (addr, self.callback_port)
+            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            try:
+                sock.bind((addr, 0))
+                sock.sendto(
+                    (SSDP_NOTIFY % callback_addr).encode("UTF-8"), ssdp_target
+                )
+            except socket.error:
+                pass
+            finally:
+                sock.close()
+
+    def respond_to_discovery(self) -> None:
+        """Respond to a WeMo discovery request with the virtual device URL."""
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        try:
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+            # Join the multicast group on all interfaces.
+            group = socket.inet_aton(MULTICAST_GROUP)
+            for addr in interface_addresses():
+                try:
+                    local = socket.inet_aton(addr)
+                    sock.setsockopt(
+                        socket.IPPROTO_IP,
+                        socket.IP_ADD_MEMBERSHIP,
+                        group + local,
+                    )
+                except socket.error:
+                    pass
+
+            sock.bind((MULTICAST_GROUP, MULTICAST_PORT))
+
+            next_notify = datetime.min
+            while not self._exit.is_set():
+                # Periodically send NOTIFY messages.
+                now = datetime.now()
+                if now > next_notify and self._notify_enabled:
+                    next_notify = now + timedelta(minutes=2)
+                    self.send_notify()
+
+                # Check for new discovery requests.
+                if not select.select([sock], [], [], 1)[0]:
+                    continue  # Timeout, no data. Loop again and check for exit.
+                msg, addr = sock.recvfrom(1024)
+                lines = msg.splitlines()
+                if len(lines) < 3 or not lines[0].startswith(
+                    b"M-SEARCH * HTTP"
+                ):
+                    continue
+                if (
+                    EXPECTED_ST_HEADER not in lines
+                    or EXPECTED_MAN_HEADER not in lines
+                ):
+                    continue
+                callback_addr = (get_ip_address(addr[0]), self.callback_port)
```

Comment:
Isn't `addr` the IP?  Is `addr[0]` just the first digit?

Reply:
`addr` here is a socket address tuple containing `(address, port)`.  I'll rename it `addr`->`sock_addr`

Reply:
ok, right, since this was created with `AF_INET`